src/array/dyn_geometry_array.rs

```diff
@@ -0,0 +1,310 @@
+    }
+
+    /// Iterator over GEOS geometry objects, taking validity into account
+    #[cfg(feature = "geos")]
+    pub fn iter_geos(
+        &self,
+    ) -> ZipValidity<geos::Geometry, impl Iterator<Item = geos::Geometry> + '_, BitIterator> {
+        ZipValidity::new_with_validity(self.iter_geos_values(), self.nulls())
+    }
+}
+
+#[inline]
+pub fn as_point_array(arr: &dyn GeometryArrayTrait) -> &PointArray {
+    arr.as_any()
+        .downcast_ref::<PointArray>()
+        .expect("Unable to downcast to point array")
+}
+
+#[inline]
+pub fn as_line_string_array<O: OffsetSizeTrait>(
+    arr: &dyn GeometryArrayTrait,
+) -> &LineStringArray<O> {
+    arr.as_any()
+        .downcast_ref::<LineStringArray<O>>()
+        .expect("Unable to downcast to line string array")
+}
+
+#[inline]
+pub fn as_polygon_array<O: OffsetSizeTrait>(arr: &dyn GeometryArrayTrait) -> &PolygonArray<O> {
+    arr.as_any()
+        .downcast_ref::<PolygonArray<O>>()
+        .expect("Unable to downcast to polygon array")
+}
+
+#[inline]
+pub fn as_multi_point_array<O: OffsetSizeTrait>(
+    arr: &dyn GeometryArrayTrait,
+) -> &MultiPointArray<O> {
+    arr.as_any()
+        .downcast_ref::<MultiPointArray<O>>()
+        .expect("Unable to downcast to multi point array")
+}
+
+#[inline]
+pub fn as_multi_line_string_array<O: OffsetSizeTrait>(
+    arr: &dyn GeometryArrayTrait,
+) -> &MultiLineStringArray<O> {
+    arr.as_any()
+        .downcast_ref::<MultiLineStringArray<O>>()
+        .expect("Unable to downcast to multi line string array")
+}
+
+#[inline]
+pub fn as_multi_polygon_array<O: OffsetSizeTrait>(
+    arr: &dyn GeometryArrayTrait,
+) -> &MultiPolygonArray<O> {
+    arr.as_any()
+        .downcast_ref::<MultiPolygonArray<O>>()
+        .expect("Unable to downcast to multi polygon array")
+}
+
+#[inline]
+pub fn as_rect_array(arr: &dyn GeometryArrayTrait) -> &RectArray {
+    arr.as_any()
+        .downcast_ref::<RectArray>()
+        .expect("Unable to downcast to rect array")
+}
+
+#[inline]
+pub fn as_wkb_array<O: OffsetSizeTrait>(arr: &dyn GeometryArrayTrait) -> &WKBArray<O> {
+    arr.as_any()
+        .downcast_ref::<WKBArray<O>>()
+        .expect("Unable to downcast to wkb array")
+}
+
+#[inline]
+pub fn as_mixed_array<O: OffsetSizeTrait>(arr: &dyn GeometryArrayTrait) -> &MixedGeometryArray<O> {
+    arr.as_any()
+        .downcast_ref::<MixedGeometryArray<O>>()
+        .expect("Unable to downcast to mixed array")
+}
+
+#[inline]
+pub fn as_geometry_collection_array<O: OffsetSizeTrait>(
+    arr: &dyn GeometryArrayTrait,
+) -> &GeometryCollectionArray<O> {
+    arr.as_any()
+        .downcast_ref::<GeometryCollectionArray<O>>()
+        .expect("Unable to downcast to geometry collection array")
+}
+
+impl TryFrom<(&Field, &dyn Array, bool)> for Box<dyn GeometryArrayTrait> {
+    type Error = GeoArrowError;
+
+    fn try_from((field, array, is_large): (&Field, &dyn Array, bool)) -> Result<Self, Self::Error> {
```

Comment:
I'm not a fan of `From` implementations on tuples like this. I find it very hard to read in practice. I think it would be clearer if it's refactored into a function.

`is_large` also shouldn't be necessary, as that can be inferred from the `field` (or the `array`).

Reply:
> is_large also shouldn't be necessary, as that can be inferred from the field (or the array).

you mean this can be inferred by arrow datatype? e.g. LargeList vs List?

Reply:
Yeah exactly